jaeger-core/src/main/java/com/uber/jaeger/samplers/ProbabilisticSampler.java

```diff
@@ -33,25 +33,38 @@
 public class ProbabilisticSampler implements Sampler {
   public static final String TYPE = "probabilistic";
 
-  private final long positiveSamplingBoundary;
-  private final long negativeSamplingBoundary;
-  private final double samplingRate;
-  private final Map<String, Object> tags;
+  private final double DELTA = 0.00001;
+  private long positiveSamplingBoundary;
+  private long negativeSamplingBoundary;
+  private double samplingRate;
+  private Map<String, Object> tags;
```

Comment:
Immutable structures are generally easier to reason about, and easier to get right, especially w.r.t. thread safety. The PerOperationSampler is inherently mutable, I am not sure what you see as being `final` there. The GuaranteedThroughputSampler is still simple enough to be immutable.

The `update` method is just one way of building an API. What you're really doing there is asking "if these were your new parameters, do you need to update?"

I still don't see the GC concerns. First, 2000 endpoints is unrealistic number - if any service has that many, they have bigger problems. But even with that number, the actual rate of change is significantly lower than once a minute, because the sampling rates are supposed to stabilize quickly.